kroxylicious-filters/kroxylicious-oauthbearer-validation/src/main/java/io/kroxylicious/proxy/filter/oauthbearer/OauthBearerValidationFilter.java

```diff
@@ -0,0 +1,212 @@
+                                    .setErrorCode(SASL_AUTHENTICATION_FAILED.code())
+                                    .setErrorMessage(e.getMessage())
+                                    .setAuthBytes(request.authBytes());
+                            LOGGER.debug("SASL Authentication failed : {}", e.getMessage(), e);
+                            return context.requestFilterResultBuilder().shortCircuitResponse(failedResponse).withCloseConnection().completed();
+                        }
+                        else {
+                            LOGGER.debug("SASL error : {}", e.getMessage(), e);
+                            return context.requestFilterResultBuilder()
+                                    .shortCircuitResponse(
+                                            new SaslAuthenticateResponseData()
+                                                    .setErrorCode(UNKNOWN_SERVER_ERROR.code())
+                                                    .setAuthBytes(request.authBytes()))
+                                    .withCloseConnection()
+                                    .completed();
+                        }
+                    });
+        }
+    }
+
+    @Override
+    public CompletionStage<ResponseFilterResult> onSaslAuthenticateResponse(short apiVersion, ResponseHeaderData header,
+                                                                            SaslAuthenticateResponseData response, FilterContext context) {
+        if (response.errorCode() == NONE.code()) {
+            this.validateAuthentication = false;
+        }
+        return context.forwardResponse(header, response);
+    }
+
+    private CompletionStage<byte[]> authenticate(byte[] authBytes) {
+        String digest;
+        try {
+            digest = digestBytes(authBytes);
+        }
+        catch (NoSuchAlgorithmException e) {
+            return CompletableFuture.failedStage(e);
+        }
+        Duration delay = strategy.getDelay(rateLimiter.get(digest).get());
+        return schedule(() -> {
+            try {
+                return CompletableFuture.completedStage(doAuthenticate(authBytes));
+            }
+            catch (Exception e) {
+                return CompletableFuture.failedStage(e);
+            }
+        }, delay)
+                .whenComplete((bytes, e) -> {
+                    if (e != null) {
+                        rateLimiter.get(digest).incrementAndGet();
+                    }
+                    else {
+                        rateLimiter.invalidate(digest);
+                    }
+                });
+    }
+
+    private byte[] doAuthenticate(byte[] authBytes) throws SaslException {
+        try {
+            byte[] bytes = this.saslServer.evaluateResponse(authBytes);
+            if (!this.saslServer.isComplete()) {
+                throw new SaslAuthenticationException("SASL failed : " + new String(bytes, StandardCharsets.UTF_8));
```

Comment:
I think we should avoid including the bytes in the error message. It's untrusted data from the client, so we can't assume it's valid utf8.

Reply:
It is the challenge response, so it is should not contain client data.

You are right that this code assumes it is an error string.  When I was testing yesterday it contained json with words 'token error' so getting this over to the client is useful for problem diagnosis. I think this is safe given this filter implementation is tied to the OauthBearerSaslServer.  String::new doesn't won't fail if the bytes aren't UTF-8.